### ISSUE TITLE
Feature Request : Add Precision, Recall and F1 Score as Stateful Metrics

### ISSUE THREAD

**Issue by SriRangaTarun** [keras-team/keras#11846]
I request that the global precision, recall and f1 metrics (for binary classification) be added to callbacks.py. It would be useful for users if they could see their model's precision, recall or f1 score on the validation data (after each epoch). Many times, these metrics are more useful than simple accuracy.
Thank you!

- [y] Check that you are up-to-date with the master branch of Keras. You can update with:
`pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps`

- [y] Check that your version of TensorFlow is up-to-date. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by Dref360** [comment]
It should be done using a StatefulMetric.

PR welcome

**Comment by SriRangaTarun** [comment]
Can someone help me explain why [this](https://github.com/keras-team/keras/pull/11851/commits/fd56ad675feb742c0db088e156d830c2c9e1279b) failed ?
Is it possible to import and use the sklearn API in callbacks.py ?

**Comment by SriRangaTarun** [comment]
Can someone please help me explain what's wrong with [this PR](https://github.com/keras-team/keras/pull/11852) ?

**Comment by SriRangaTarun** [comment]
Can someone please tell me what could be wrong with this code ? I added it to the callbacks.py class.

class BinaryClassificationMetrics(Callback):
    def on_train_begin(self, logs=None):
        self.val_precisions = []
        self.val_f1s = []
        self.val_recalls = []
        
    def on_epoch_end(self, epoch, logs=None):
        logs = logs or {}
        val_predict = (np.asarray(self.model.predict(self.model.validation_data[0]))).round()
        val_targ = self.model.validation_data[1]
        _val_precision = np.dot(val_targ, val_predict)/sum(val_predict)
        _val_recall = np.dot(val_targ, val_predict)/sum(val_targ)
        _val_f1 = (2*_val_precision*_val_recall)/(_val_precision + _val_recall)
        self.val_precisions.append(_val_precision)
        self.val_recalls.append(_val_recall)
        self.val_f1s.append(_val_f1)
        print(" — val_f1: %f — val_precision: %f — val_recall %f" %(_val_f1, _val_precision, _val_recall))
        return  

class TerminateOnNaN(Callback):
    """Callback that terminates training when a NaN loss is encountered.
    """

    def on_batch_end(self, batch, logs=None):
        logs = logs or {}
        loss = logs.get('loss')
        if loss is not None:
            if np.isnan(loss) or np.isinf(loss):
                print('Batch %d: Invalid loss, terminating training' % (batch))
                self.model.stop_training = True

**Comment by shreyashub** [comment]
Is the issue still open?

**Comment by oasis789** [comment]
If this issue is still open, can I pick it up and work on it? I'm new to the world of open source and want to work on something relatively easy for my first contribution.

**Comment by adam2392** [comment]
If this issue is still open, I would be interested in adding additional functionality to this where performance metrics are put in a dictionary format at varying levels of precision like 100, 90, 80, 50%. Very helpful for not optimizing accuracy necessarily, but potentially controlling number of false positives. 

I essentially have to hack it currently.

**Comment by andy6975** [comment]
Hi, if this issue is still open, I would like to work on it. Thanks!

**Comment by erichulburd** [comment]
It seems to me that Recall and Precision can already be passed as stateful metrics as follows (see example.ipynb for complete code):

```
model.compile(optimizer=optimizers.RMSprop(lr=0.001),
              loss=losses.binary_crossentropy,
              metrics=[metrics.binary_accuracy, metrics.Recall(), metrics.Precision(), metrics.F1Score()])
```

However, F1 score is not implemented. I made a naive approach at implementing it here: https://github.com/erichulburd/keras/commit/bc18a9cc35dcd8c8804cd1c193b9d149d2d3ce29

Questions:

- Is this worth pursuing as a PR?
- Should I store recall and precision in the `ConfusionMatrix` enum to avoid repeated operations?
- Should I abstract Recall, Precision, and F1Score in metrics.py to reduce duplicate code?

**Comment by MrinalTyagi** [comment]
I have tried to solve this issue and created a binary classifier callback that can help in outputting the validation precision recall and f1 score along with the already provided loss and accuracy. The output of the following code looks like following =>
![keras](https://user-images.githubusercontent.com/21031150/121186601-e6e5f300-c884-11eb-854e-9902373a440d.png)

Here is the colab link containing the callback class [here](https://colab.research.google.com/drive/1EZSreQFRUvUjAIWCuTZoi2JhV-MgxP45?usp=sharing)

@Dref360 Could you go through the changes and let me know if it is worth a PR?
